timemachine/fe/free_energy.py

```diff
@@ -898,6 +904,61 @@ def compute_intermediate_result(lambdas: Sequence[float]) -> PairBarResult:
     return results, trajectories
 
 
+def compute_potential_matrix(
+    potential: custom_ops.Potential,
+    hrex: HREX[CoordsVelBox],
+    params_by_state: NDArray,
+    n_neighbor_states: Optional[int] = None,
+) -> NDArray:
+    """Computes the (n_replicas, n_states) sparse matrix of potential energies, where a given element $(k, l)$ is
+    computed iff state $l$ is within `n_neighbor_states` of the current state of replica $k$, and is otherwise set to
+    `np.inf`.
+
+    Parameters
+    ----------
+    potential : custom_ops.Potential
+        potential to evaluate
+
+    hrex : HREX
+        HREX state (containing replica states and permutation)
+
+    params_by_state : NDArray
+        (n_states, ...) array of potential parameters for each state
+
+    n_neighbor_states : int or None, optional
+        If given, number of neighbor states on either side of a given replica's state for which to compute potentials.
+        If None, compute energies for all (replica, state) pairs.
+    """
+
+    coords = np.array([xvb.coords for xvb in hrex.replicas])
+    boxes = np.array([xvb.box for xvb in hrex.replicas])
+
+    def compute_sparse(k: int):
+        n_states = len(hrex.replicas)
+        state_idx = np.argsort(hrex.replica_idx_by_state)
+        neighbor_state_idxs = state_idx[:, None] + np.arange(-k, k + 1)[None, :]
+        valid_idxs = np.nonzero((0 <= neighbor_state_idxs) & (neighbor_state_idxs < n_states))
+        coords_batch_idxs = valid_idxs[0].astype(np.uint32)
```

Comment:
Are the coordinates sorted such that a series of parameter evaluations can be done without a neighborlist rebuild? Not clear to me from the code, but if not then might explain why there isn't a performance improvement. Or perhaps the cost of recomputing energies with different parameters is negligible compared to neighborlist rebuilds (though less confident of that)

Reply:
Hmm, actually it turns out that the ordering happens to already be what I'd imagine is optimal for avoiding neighborlist rebuilds (i.e. outer loop over coordinates in replica order and inner loop over parameters), so I'm not sure if this is the issue after all. Will keep investigating.

Reply:
After more careful [benchmarking](#benchmarks) with 3 replicates, the performance boost does seem mostly as expected. I think I was not careful enough with the initial benchmarks (for instance might have run other processes concurrently while the benchmarks were running).

One feature I'm not sure I understand yet is why, for solvent specifically, the ns_per_day measurement does not flatten out for larger number of windows (though it's still a significant improvement).